Fix dbWriteTable and temporary tables, #10

gaborcsardi opened the pull request:
Also, `append = TRUE` now does not create the table if it does not exist, but returns an error. These two changes were discussed together in #10, but they are actually independent, imo. The new `append = TRUE` behavior might break existing code that relies on the current behavior, so maybe we don't want that.

As for the temporary table part, the only sane way to check if a temporary or regular table exists, is to run a `SELECT` on it, seemingly. :( This might be problematic if the user has privs to run `SHOW TABLES`, but not the `SELECT`. This does not happen very often I guess, and here we are trying to write to a table, anyway, although that is again a different priv.

What to you think? This is probably too much ado about too little, but wanted to "finish" it, nevertheless.

hadley commented:
Could you please merge/rebase? Other comments in line

hadley commented on R/table.R:
```diff
@@ -203,6 +210,16 @@ setMethod("dbExistsTable", c("MySQLConnection", "character"),
   }
 )
 
+dbExistsTableOrTemporaryTable <- function(conn, name) {
```
I'd rather make this the default for dbExists

gaborcsardi replied:
You mean `dbExistsTable`?

hadley commented on R/table.R:
```diff
@@ -203,6 +210,16 @@ setMethod("dbExistsTable", c("MySQLConnection", "character"),
+  try_res <- try(
+    silent = TRUE,
+    dbGetQuery(conn, paste0(
+      "SELECT * FROM ", dbQuoteIdentifier(conn, name), " WHERE FALSE"
```
Maybe "SELECT NULL from" ...

gaborcsardi replied:
OK

hadley commented on R/table.R:
```diff
@@ -203,6 +210,16 @@ setMethod("dbExistsTable", c("MySQLConnection", "character"),
 )
 
+dbExistsTableOrTemporaryTable <- function(conn, name) {
+  try_res <- try(
```
I think a better pattern for this sort of try is 

```R
tryCatch({
  dbGetQuery(...)
  TRUE
}, error = function(...) {
  FALSE
})
```

gaborcsardi replied:
OK.

hadley commented on tests/testthat/test-dbWriteTable.R:
```diff
@@ -15,3 +15,54 @@ test_that("options work", {
+test_that("appending is error if table does not exist", {
+  if (!mysqlHasDefault()) skip("Test database not available")
+
+  con <- dbConnect(MySQL(), dbname = "test")
```
Can you please switch to the new style `con <- mysqlDefault()` (and drop the explicit `skip()`)

gaborcsardi replied:
OK.

hadley commented on tests/testthat/test-dbWriteTable.R:
```diff
@@ -15,3 +15,54 @@ test_that("options work", {
+  con <- dbConnect(MySQL(), dbname = "test")
+  on.exit(dbDisconnect(con))
+
+  dbGetQuery(
```
You can now use the `temporary = TRUE` argument to `dbWriteTable()`

gaborcsardi replied:
OK, but then if there is a `temporary` argument, then my PR does not make much sense.  Btw. even then, I would not change this test, because here I test how dbWriteTable behaves with an already existing temporary table.

gaborcsardi replied:
But actually it does not matter much how the existing table was created. So why not? Done.

gaborcsardi commented:
OK, updated all and rebased. (Btw. I get some segfaults sometimes when testing, but these were the same before, so most likely separate issue. Will report it if I'll have time.)

hadley commented:
Thanks!
